### ISSUE TITLE
Sort does not work if using valuePrepareFunction [ISSUE]

### ISSUE THREAD

**Issue by karangarg45** [akveo/ng2-smart-table#296]
if i am using valuePrepareFunction on any column then sorting doesnot work on any of the columns.Could you please check?

**Comment by hgarc014** [comment]
I'm using version `0.5.0-6` and I have this issue

**Comment by karangarg45** [comment]
the solution is to write custom sort function.it does not work because it tries to sort on the raw value instead of the value passed from the value prepare function.

**Comment by hgarc014** [comment]
Yes, I just figured this out. I believe you also need to update the filter field if you want to do filter on the column as well

**Comment by karangarg45** [comment]
yes thats true.

**Comment by steve-frank** [comment]
hi did you get a solution for this, can  you pls share the snippet?

**Comment by karangarg45** [comment]
the solution is to write custom sort function.it does not work because it tries to sort on the raw value instead of the value passed from the value prepare function.

**Comment by jagabs** [comment]
@karangarg45  hi, I tried to make a custom sort but i get empty object with cell in my 
filterFunction(cell?: any, search?: string): boolean {

**Comment by mariummalik22** [comment]
@karangarg45 Can you please give an example? Where do I define that it has to take value from my valuePrepareFunction() ?

**Comment by jagabs** [comment]
empty in console but the value is showed in the table when returned weird

**Comment by mariummalik22** [comment]
@jagabs Is your sorting working? Can you share your code?

**Comment by rami-alloush** [comment]
I had this problem when using `` to get the `count` of my data. The default sort tries to sort based on the object not it's length. I implemented this function to solve the problem

```
export function sortBasedOnCount(direction: any, a: any, b: any) {
    // Get the length of the full object
    const first = a ? Object.keys(a).length : 0;
    const second = b ? Object.keys(b).length : 0;
    if (first < second) { return -1 * direction; }
    if (first > second) { return direction; }
    return 0;
}
```
and I call it inside the desired column
```
                   ...
                    valuePrepareFunction: (value) => Object.keys(value).length,
                    compareFunction: sortBasedOnCount,
                   ...
```
